client/redfish.c

```diff
@@ -0,0 +1,3097 @@
+
+static ni_bool_t
+ni_smbios_decode_strings(ni_string_array_t *strings, ni_buffer_t *bp)
+{
+	char *beg, *end;
+
+	if (!(beg = ni_buffer_pull_head(bp, 1)))
+		return FALSE;
+	if (!(end = ni_buffer_pull_head(bp, 1)))
+		return FALSE;
+
+	while (beg[0] || end[0]) {
+		if (!end[0] && beg[0]) {
+			ni_smbios_string_array_append(strings, beg);
+			beg = end;
+		} else if (!beg[0])
+			beg = end;
+
+		if (!(end = ni_buffer_pull_head(bp, 1)))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static ni_bool_t
+ni_smbios_decode_entry(ni_smbios_decoder_t *decoder,
+			const ni_smbios_epi_t *epi,
+			const ni_string_array_t *strings,
+			const ni_smbios_header_t *header,
+			ni_buffer_t *buffer)
+{
+	const ni_smbios_handler_t *handler;
+	unsigned int i;
+
+	ni_debug_verbose(NI_LOG_DEBUG3, NI_TRACE_READWRITE,
+			"Handle 0x%04x, DMI type %u, %u bytes",
+			header->handle, header->type, header->length);
+
+	for (i = 0; i < strings->count; ++i) {
+		const char *str = strings->data[i];
+		ni_debug_verbose(NI_LOG_DEBUG3, NI_TRACE_READWRITE,
+				"	String[%u]: `%s`", i, str);
+	}
+
+	for (handler = decoder->handler; handler->decode; ++handler) {
+		if (handler->type != header->type)
+			continue;
+
+		if (!handler->decode(decoder, epi, strings, header, buffer)) {
+			ni_debug_verbose(NI_LOG_DEBUG3, NI_TRACE_READWRITE,
+					"Unable to decode DMI type %u, Handle %04x",
+					header->type, header->handle);
+		}
+	}
+
+	return TRUE;
+}
+
+static ni_bool_t
+ni_smbios_decode_tables(ni_smbios_decoder_t *decoder, const ni_smbios_epi_t *epi,
+			ni_buffer_t *tab)
+{
+	ni_string_array_t strings = NI_STRING_ARRAY_INIT;
+	const ni_smbios_header_t *header;
+	ni_buffer_t buffer;
+	unsigned char *ptr;
+	unsigned int num = 0;
+
+	while ((header = ni_buffer_peek_head(tab, sizeof(*header)))) {
+		/* Entry structure length must include header length  */
+		if (header->length < sizeof(*header))
+			break;
+
+		/* Stop at End of Table type or if numEntries reached */
+		if (header->type == 0x7f) {
+			ni_debug_verbose(NI_LOG_DEBUG3, NI_TRACE_READWRITE,
+				"Handle 0x%04x, DMI type %u, %u bytes",
+				header->handle, header->type, header->length);
+		}
+		if (header->type == 0x7f ||
+		    (epi->ent_num && ++num > epi->ent_num)) {
+			ni_debug_verbose(NI_LOG_DEBUG3, NI_TRACE_READWRITE,
+				"End Of Table");
+			break;
+		}
+
+		/* Extract type structure buffer including header */
+		if (!(ptr = ni_buffer_pull_head(tab, header->length)))
+			break;
+		ni_buffer_init_reader(&buffer, ptr, header->length);
+
+		/* Entry [strings +] structure terminator */
+		ni_smbios_string_array_reset(&strings);
+		if (!ni_smbios_decode_strings(&strings, tab))
+			break;
+
+		if (!ni_smbios_decode_entry(decoder, epi, &strings, header, &buffer))
+			break;
+	}
+
+	ni_smbios_string_array_destroy(&strings);
+	return TRUE;
+}
+
+static void
+ni_smbios_entry_free(ni_smbios_entry_t *entry)
+{
+	if (entry) {
+		if (entry->free)
+			entry->free(entry);
+		else
+			free(entry);
+		memset(entry, 0, sizeof(*entry));
```

Comment:
Writing on buffer which just was free'd isn't that nice. Maybe you will zero the memory before `free()`. The free-callback should do the job as it wants!

```suggestion
		else {
			memset(entry, 0, sizeof(*entry));
			free(entry);
		}

```